tools/integration_tests/concurrent_operations/high_cpu_concurrent_listing_test.go

```diff
@@ -0,0 +1,522 @@
+// the degree of parallelism. By default it uses GOMAXPROCS.
+// Ref: https://stackoverflow.com/questions/24375966/does-go-test-run-unit-tests-concurrently
+type highCpuConcurrentListingTest struct{}
+
+func (s *highCpuConcurrentListingTest) Setup(t *testing.T) {
+	testDirPath = setup.SetupTestDirectory(testDirName)
+}
+
+func (s *highCpuConcurrentListingTest) Teardown(t *testing.T) {}
+
+func createDirectoryStructureForTestCaseParallel(t *testing.T, testCaseDir string) {
+	operations.CreateDirectory(path.Join(testDirPath, testCaseDir), t)
+
+	explicitDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	operations.CreateDirectory(explicitDir, t)
+	numFiles := 20
+	nestedLevel := 6
+
+	var globalWG sync.WaitGroup
+
+	createFilesInGivenDir := func(dir string) {
+		var wg sync.WaitGroup
+		for i := 0; i < numFiles; i++ {
+			wg.Add(1)
+			go func(i int) {
+				defer wg.Done()
+				fileName := fmt.Sprintf("file%d.txt", i+1)
+				operations.CreateFileWithContent(path.Join(dir, fileName), setup.FilePermission_0600, "test_content", t)
+			}(i)
+		}
+
+		wg.Wait()
+	}
+
+	lastLevel := explicitDir
+	for level := 0; level < nestedLevel; level++ {
+		currLevel := path.Join(lastLevel, fmt.Sprintf("level%d", level+1))
+		lastLevel = currLevel
+		operations.CreateDirectory(currLevel, t)
+		globalWG.Add(1)
+		go func() {
+			defer globalWG.Done()
+			createFilesInGivenDir(currLevel)
+		}()
+	}
+	globalWG.Wait()
+}
+
+func listDirectoryRecursivelyWithCmd(t *testing.T, root string) {
+	cmd := exec.Command("ls", "-R", root)
+	_, err := cmd.CombinedOutput()
+
+	if err != nil {
+		t.Errorf("Error in listing recursively: %v", err)
+	}
+}
+
+////////////////////////////////////////////////////////////////////////
+// Test scenarios
+////////////////////////////////////////////////////////////////////////
+
+// Test_RecursiveListing tests for potential deadlocks or race conditions
+// when multiple goroutines performs recursive listing.
+func (s *highCpuConcurrentListingTest) Test_RecursiveListing(t *testing.T) {
+	if runtime.NumCPU() < requiredCPUCoresToRunThisTest {
+		t.SkipNow()
+	}
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_RecursiveListing"
+	createDirectoryStructureForTestCaseParallel(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	availableCPU := runtime.NumCPU() / 2 // Keep half for gcsfuse process.
+	t.Logf("Testing with %d go-routine: ", availableCPU)
+	goRoutineCountPerOperation := availableCPU // Use all cpus for recursive listing.
+	timeout := 600 * time.Second
+
+	// Create multiple go routines to listing concurrently.
+	for r := 0; r < goRoutineCountPerOperation; r++ {
+		wg.Add(1)
+		// Repeatedly do recursive listing.
+		go func() {
+			defer wg.Done()
+			for j := 0; j < iterationsForMediumOperations; j++ {
+				listDirectoryRecursivelyWithCmd(t, targetDir)
+			}
+		}()
+	}
+
+	// Wait for goroutines or timeout
+	done := make(chan bool, 1)
```

Comment:
I suppose a channel with capacity 0 should suffice here.
```suggestion
	done := make(chan bool)
```